csharp-selenium-webdriver-sample/azure-pipelines.yml

```diff
@@ -1,22 +1,55 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
+# This file defines an Azure Pipelines build, which you can configure on https://dev.azure.com
+# to run automatically with each Pull Request update and/or each commit to your repository.
+# In this example, we do not change the default "trigger" behavior, so this build gets run
+# for all commits/PRs to all branches of the axe-pipelines-samples repository.
+#
+# If you already have an Azure Pipelines build for your project, we recommend integrating
+# your new accessibility tests into it, rather than creating a separate pipeline. Pipelines
+# builds are usually defined in files like this one named "azure-pipelines.yml" or "build.yaml",
+# but some projects use different file names.
+#
+# If you are using an Azure Pipelines build with tasks configured using the GUI on https://dev.azure.com,
+# instead of a checked in YAML file, you can add the same tasks listed below from the GUI. However, we
+# strongly recommend switching to using a checked in YAML file.
+#
+# Most C# projects with an Azure Pipelines build will already have most of the steps shown below
+#
+# If you *don't* already have an Azure Pipelines build, we recommend following Azure Pipelines'
+# documentation at https://docs.microsoft.com/en-us/azure/devops/pipelines/create-first-pipeline
+# (using this sample as your starter code).
+#
+# For more information, see:
+# * https://docs.microsoft.com/en-us/azure/devops/pipelines/get-started
+# * https://docs.microsoft.com/en-us/azure/devops/pipelines/customize-pipeline
+
 pool:
+  # The 'windows-2019' and 'vs20xx-windows' vmImages come pre-installed with selenium drivers for common browsers.
```

Comment:
why would selenium drivers be installed on vms? shouldnt this be part of the test libs? I also see this for the typescript project.

Reply:
If you look at the 2 links in the comments
 - https://docs.microsoft.com/en-us/azure/devops/pipelines/test/continuous-test-selenium
 - https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2019-Server2019-Readme.md
It's mentioned in them that Selenium is pre-installed on any windows image
why is it pre-installed? 
- I think so it'd be compatible with the browser versions installed on the same agent.
- The paths to the folders containing these drivers can be obtained from the environment variables named IEWebDriver (Internet Explorer), ChromeWebDriver (Google Chrome), and GeckoWebDriver (Firefox).
- That's why we do something like: var chromeDriverDirectory = Environment.GetEnvironmentVariable("ChromeWebDriver") ?? Environment.CurrentDirectory;
- So it should works fine on windows agents, other agents and local ,

Reply:
I agree that ideally we'd pin versions of all the test dependencies. The issue is that:

* there isn't a particularly simple/official way to pull down and install pinned versions of the actual browsers in a cross-platform manner; in particular, Google doesn't seem to offer officially supported channels for downloading non-current versions of Chrome.
* The webdrivers (particularly ChromeDriver) very explicitly only support being used with exactly-matching browser versions

The result is that the webdriver and the browser have to be kept in sync. When we debated how to do this in the typescript sample, we felt the simplest method for this sample to use was to use the synced webdriver/browser pair already present on the agents; since this sample is intended to be primarily about integrating axe, rather than setting up Selenium, we traded off the more robust version pinning for simplicity.